### PR TITLE
Initial checkout block support

### LINKED ISSUE #652: Support the checkout block
Related to adding WCPay to https://github.com/woocommerce/woocommerce-gutenberg-products-block and #639

cc @thenbrent  @bmccotter

### PR DESCRIPTION
**Pull request by RadoslavGeorgiev**
Fixes #652

### Changes proposed in this Pull Request

This PR was supposed to be a spike, related to Stripe's React elements, but grew into a PR with support for WooCommerce Blocks. 😲 

This was done by following https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/docs/extensibility/payment-method-integration.md for the checkout block, and isolating some functionality from the existing `wcpay-checkout.js`.

### Testing instructions

1. Make sure that the classic checkout is not broken. This includes:
    - Testing with both a simple card (`**** 4242`) and an SCA-enabled card (ex. `4000002760003184`).
    - Going through a successful workflow.
    - Checking whether all error messages (or failure scenarios) are handled in the correct location on the page (you'll know it ;) ).
2. Install https://github.com/woocommerce/woocommerce-gutenberg-products-block and create a page with it.
3. Retry all of the tests from the classic checkout within the checkout block.

#### All scripts and dependencies are loaded within Gutenberg

Is this expected, or it should somehow happen dynamically? Even when composing a new post without a checkout block, Stripe's API and all necessary scripts are loaded.

#### Erroneous state after confirmation

error-notice

If there is an error during 3DS authentication (through the `onCheckoutAfterProcessingWithSuccess` handler):

1. The error message looks broken. The X button is misaligned.
2. The page does not scroll to the error message, so it remains above the fold.
3. Even though the confirmation failed, the submit button displays a seemingly successful check-mark.<br />fields-after-error

#### Webpack/Babel issues between checkout method

Everything, related to the new checkout (with blocks) gets compiled properly. However, trying to use `async` functions for the classic checkout fails.

Any idea why this might be happening? Both `client/checkout/blocks/index.js` and `client/checkout/classic/index.js` are loaded the same way in webpack's config.

P.S. There is a failing PHPUnit test, which I will look into, but is not blocking the review of the PR.

### CODE REVIEW

File: client/checkout/blocks/index.js
Comment (haszari) [inline client/checkout/blocks/index.js:29]: Payment methods can show an icon – might be nice to have Woo Pay icon on the tab. See example in [Stripe CC](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/assets/js/payment-method-extensions/payment-methods/stripe/credit-card/index.js#L41) or cheque. 

Screen Shot 2020-07-10 at 4 45 22 PM

File: client/checkout/blocks/index.js
Comment (haszari) [inline client/checkout/blocks/index.js:22]: Should there be a version check around this? E.g. old version of WooCommerce/blocks that doesn't have blocks payment methods API. (Maybe there's already a higher-level version check in this plugin somewhere?)

**Comment by haszari** [comment]
I've had a quick look over the code and don't see any issues. I also did not test, but I'm happy to give this a good test (need to get a local build going) - let me know if you'd like me to test :)

**Comment by haszari** [comment]
I have tested this in new `Checkout` block locally (with `DEV_MODE`) and successfully paid for an order!

Screen Shot 2020-07-15 at 8 53 44 AM

Screen Shot 2020-07-15 at 8 55 35 AM

I'll do some more testing and a review soon, but just wanted to 🎉 that this is working with WooCommerce Blocks!

File: client/checkout/classic/index.js
Comment (haszari) [reply client/checkout/classic/index.js:10]: FYI - I didn't test classic checkout either, happy to do more testing if you like - ping me :)

File: client/checkout/blocks/index.js
Comment (nerrad) [reply client/checkout/blocks/index.js:27]: For reference, you can see the tweak we made to the API [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2840)

File: client/checkout/blocks/index.js
Comment (nerrad) [reply client/checkout/blocks/index.js:22]: I don't think this file will get loaded in the client if it's not registered on the server and it won't get registered on the server if the action server side is not available. So I think they should be good here.

File: client/checkout/blocks/index.js
Comment (LevinMedia) [reply client/checkout/blocks/index.js:29]: We're removing payment method icons, unless it's a consumer brand, such as paypal - so WC pay won't need a logo. 👍

File: client/checkout/blocks/fields.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/fields.js:42]: The anonymous function _will_ return the unsubscription callback because of the lack of `{}`s. Is there something I'm missing?

File: client/checkout/blocks/confirm-card-payment.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/confirm-card-payment.js:31]: Thank you, I changed this in 5a49f3b and it works.

> With some refactoring of Woo Core we could indicate what error types are the cause for errors and use that to divvy up where notices appear but for now it's best to allow payment methods to indicate where they want their error notices to appear.

I agree that this is probably the best way to do it for now.

File: client/checkout/blocks/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/index.js:27]: This looks very promising and I'll keep it in mind for future development, however I'm not sure whether it's useful to the current implementation.

The availability of the gateway is already determined in [PHP](https://github.com/Automattic/woocommerce-payments/pull/723/files#diff-b884312edfda620543d7d8c2222df36aR27). As long as that works and API keys are present, it should be all fine.

In 03483be I modified the method to check whether the Stripe JS is loaded, but I'm not even sure if that makes any sense.

File: client/checkout/api/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/api/index.js:71]: True, the name was misleading 😅 

I renamed the method to `prepareValue` in 754463e.

File: client/checkout/api/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/api/index.js]: We'll discuss the plans in paJDYF-M7-p2, however for the moment I changed things in d624d27 in order to avoid jQuery in the blocks checkout.

File: client/checkout/blocks/fields.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/fields.js:20]: I'm ready to change the name to a better one, but I'm not sure `WCPayForm` is it, since it's not an actual form, but payment fields.

File: client/checkout/blocks/fields.js
Comment (nerrad) [reply client/checkout/blocks/fields.js:43]: 🤦 Oh right! I missed that. Move along, nothing to see here... 😄

File: client/checkout/blocks/fields.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/fields.js:34]: Good point, implemented in ecf25e7 👍

File: client/checkout/blocks/fields.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/fields.js:51]: I think the complication comes from the fact that it's called element/elements. This is a Stripe idiom, regarding to Stripe Elements, which are neither HTML nor React elements.

I could use the `onReady` prop of `CardElement` to access the element, but I'd need to save it in the state and that would be very confusing. I'm also pretty sure (not sure, but pretty sure) that `getElement` does not traverse the DOM, but loads an initialized element instance of the given type.

File: client/checkout/blocks/fields.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/fields.js:84]: I agree, but <CardElement /> comes from Stripe's React library and as you can see in https://stripe.com/docs/stripe-js/react#element-props, that's the way it works.

The reason is that `options` are passed to the underlying Elements instance, while the other props are used by their React library.

File: client/checkout/blocks/generate-payment-method.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/generate-payment-method.js:27]: It is neither a test request, nor the payment request. Here is where Stripe generates a tokenized payment method based on the fields on the page, hence the name `generatePaymentMethod`. 

I renamed the function (in the API class) to `generatePaymentMethodRequest` in bc06347, hopefully this makes it clearer.

File: client/checkout/classic/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/classic/index.js:10]: The classic checkout should be mostly not affected by the changes here. Once I get a 👍 from you, I'll request a review from another Hydra member.

File: client/checkout/classic/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/classic/index.js:19]: Generally the script should not be enqueued without the checkout form, however I agree with you. I added a check whether the configuration is available in 271d448, so that no JS errors would be present.

File: client/checkout/blocks/generate-payment-method.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/generate-payment-method.js:11]: Added in ac4bad2 +1

File: client/checkout/api/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/api/index.js]: As noted in https://github.com/Automattic/woocommerce-payments/pull/723#discussion_r459489908, the method is not renamed to `generatePaymentMethodRequest`. By the way, "Payment Method" comes from Stripe's terminology and I use it intentionally, as the previous tokenized card API is called `sources`, and even used to be `cards` before.

> What are the benefits of using a class for this vs more functional approach?

A functional approach would work as well, but I did it this way in order to make it clear that some methods (e.g. `prepareValue`, `setBillingDetail`, and `setAddressDetail`) are specific to this request. I find this approach more readable and cleaner than the way it used to be in the old checkout file.

**Comment by RadoslavGeorgiev** [comment]
@nerrad the PR is nearly ready for another review round, but I'm having an issue.

Please look at @haszari's comment above. The gateway throws errors even when not selected. To avoid this, I wanted to use the `activePaymentMethod` prop, but it never changes. If `cheque` is selected initially, `activePaymentMethod` remains `cheque` even when WCPay is selected. Any idea why?

**Comment by nerrad** [comment]
> The gateway throws errors even when not selected. To avoid this, I wanted to use the activePaymentMethod prop, but it never changes. If cheque is selected initially, activePaymentMethod remains cheque even when WCPay is selected. Any idea why?

I'm not able to reproduce this with the latest built blocks code in our repo (just with our current payment methods, not with this branch or wcPay installed).  If indeed the `activePaymentMethod` prop is not updating for you, then I figured it should be reproducible in the blocks repo alone. If I add some console.logging here:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/9351832c11a843d34572bc717df20d315254e2b0/assets/js/payment-method-extensions/payment-methods/stripe/credit-card/payment-method.js#L86-L88

I find that selecting different payment methods in the tab works as expected.  Where did you try using `activePaymentMethod` from props? If you try adding it [around here](https://github.com/Automattic/woocommerce-payments/blob/ac4bad20d8c3b59ffc4b05bf1a577a81559273a9/client/checkout/blocks/fields.js#L87-L90) (similar to what I did for stripe cc) does that work?

As for a review, either myself or Rua will do another review (hopefully before end of week).

File: client/checkout/blocks/index.js
Comment (haszari) [reply client/checkout/blocks/index.js]: Sounds good! Are there any client restrictions/limitations for WC Pay, e.g. geo, browser? If so, they might be appropriate criteria for `canMakePayment`.

File: client/checkout/classic/index.js
Comment (haszari) [inline client/checkout/classic/index.js:111]: Looks like this needs updating (now `generatePaymentMethodRequest`?). I get `generatePaymentMethodRequest is not a function` error when using WC Pay in classic (shortcode) checkout.

File: client/checkout/blocks/fields.js
Comment (haszari) [inline client/checkout/blocks/fields.js:88]: As suggested by @nerrad, you'll likely want to check `props.activePaymentMethod` here to ensure that Stripe/WC Pay is not invoked when shopper uses a different gateway. 

I tested it with the following change:

```js
const ConsumableWCPayFields = ( { api, ...props } ) => {
	const { activePaymentMethod } = props;
	return ( activePaymentMethod === PAYMENT_METHOD_NAME ) && (
		<Elements stripe={ api.getStripe() }>
			<ElementsConsumer>
				{ ( { elements, stripe } ) => (
					<WCPayFields api={ api } elements={ elements } stripe={ stripe } { ...props } />
				) }
			</ElementsConsumer>
		</Elements>
	);
}
```

This fixes the issue I saw in testing - if WC Pay is activated as a gateway, and shopper selects a different gateway (e.g. `Check`), the Stripe elements form validation fires, showing a spurious "Please enter a valid credit card number" error :)

File: client/checkout/classic/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/classic/index.js]: Oh, my bad. I renamed the method and one of the instances where it's used, but not the other one. Fixed in ce8919d

File: client/checkout/blocks/fields.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/fields.js]: Good point 👌 , fixed in 0ea9d79

File: client/checkout/blocks/fields.js
Comment (RadoslavGeorgiev) [reply client/checkout/blocks/fields.js:94]: Actually, my mistake was that I was trying to use `activePaymentMethod` in the effect, without indicating the dependency on it. https://github.com/Automattic/woocommerce-payments/pull/723/commits/9204ccd9fe84e1eb350799485d5c327f5e32e059 fixes this.

Your method would work when switching from Cheque to WCPay, but not when swiching back to Cheque, because the component would be already initialized.

**Comment by RadoslavGeorgiev** [comment]
@nerrad and @haszari I'm done with all of the requested changes. You can take a look, but I'll also add the PR to the Hydra review queue.

> If you like, I'm happy to do some more testing on an ephemeral site (https/real domain) with released version of woo blocks (once the two issues above are sorted, and you/hydra are done with the PR).

Unless either of you has access to a 3DS2-enabled card, I don't think it's necessary.

**Comment by haszari** [comment]
Nice work @RadoslavGeorgiev !

> > If you like, I'm happy to do some more testing on an ephemeral site (https/real domain) with released version of woo blocks (once the two issues above are sorted, and you/hydra are done with the PR).

> Unless either of you has access to a 3DS2-enabled card, I don't think it's necessary.

I think I do have such a card (I get a confirm modal on eCommerce sites) - though not sure if it's 3DS specifically. Hit me if I can help with further testing :)

**Comment by dwainm** [comment]
@RadoslavGeorgiev could we push the latest master to this? I'd like to test it with all the latest changes.

**Comment by RadoslavGeorgiev** [comment]
@dwainm I merged the latest changes from `master` into this branch.

In order to avoid additions to this PR, since it's already massive, I have not added any subscription- or saved cards-specific code. I tested that using saved cards works, but there is no option to save new cards.

**Comment by DanReyLop** [comment]
@RadoslavGeorgiev The fix for IE11 was pretty much the same as what you implemented in this PR (adding the "classic" checkout to the Webpack pipeline), so there are some conflicts with `master`. I've solved the conflicts and re-applied the changes[ in the `spike/blocks-checkout-merge-with-ie11-fix` branch](https://github.com/Automattic/woocommerce-payments/compare/spike/stripe-react-elements...spike/blocks-checkout-merge-with-ie11-fix?expand=1).

**Review by dwainm — APPROVED** [review]
Works, but had difficulty testing SCA, but I think this is good to go since we'r targeting initial support.

I check that that old checkout still works.

Lastly saving new cards is not enabled and I think we should add an issue for this.

**Comment by DanReyLop** [comment]
> but had difficulty testing SCA

@dwainm What do you mean by that? SCA doesn't work, or you couldn't test it? It's not a blocker for this PR either way, but if it doesn't work we should open a follow-up issue so we don't forget about it.

**Comment by vbelolapotkov** [comment]
@RadoslavGeorgiev I've noticed e2e test is failing for saved cards with this change applied. I haven't checked deeper yet, and wanted to ask first could PR changes cause this?

Screenshot is here p1600774926020400-slack-CQ0Q6N62D.

**Comment by dwainm** [comment]
@DanReyLop I had significant issues with the Jetpack autoloader and some of the timeout issues we were having on the server, so even though I tried the SCA popup didn't show for me. I gave up trying and mentioned it as @RadoslavGeorgiev was able to get it working and therefore it should not be a blocker.

File: client/checkout/api/index.js
Comment (allendav) [inline client/checkout/api/index.js:147]: Even after enabling create-account-during-checkout in settings, I didn't see such an option on the block based checkout. How does one enable this?

File: client/checkout/api/index.js
Comment (allendav) [reply client/checkout/api/index.js:173]: ^^ @RadoslavGeorgiev

File: client/checkout/classic/index.js
Comment (allendav) [inline client/checkout/classic/index.js:277]: StackOverflow has a shorthand URL that might be clearer here:

https://stackoverflow.com/a/5298684

File: client/checkout/api/index.js
Comment (haszari) [reply client/checkout/api/index.js:147]: We're working on it! https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2377 https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2851

See also P2 – supporting signup has triggered iteration/work in a few related areas (and lots of discussion). cc @nerrad

File: includes/class-wc-payment-gateway-wcpay.php
Comment (allendav) [inline includes/class-wc-payment-gateway-wcpay.php:286]: By "check your email" do you mean their email inbox or do you mean the address on the screen. (I think you mean the former.) Please add "inbox" or "address" as appropriate after email.

File: includes/class-wc-payments-payment-method.php
Comment (allendav) [inline includes/class-wc-payments-payment-method.php:13]: This class name should probably mention Blocks, no? Otherwise it might be unclear to maintainers that this is Blocks specific code.

File: includes/class-wc-payments.php
Comment (allendav) [inline includes/class-wc-payments.php:491]: Similar to comment above, please consider including "blocks" in the name of this file

**Review by allendav — CHANGES_REQUESTED** [review]
Tested with a DreamHost based fresh install of WordPress, WooCommerce, Storefront and WooCommerce Blocks. Successfully tested with logged in user and logged out.

Could not figure out how to create account during logged out checkout.

Added a few comments and boosted one of @nerrad 's that might have been overlooked.

Did not run unit tests. Please fix conflict, rebase and let's see how Travis does.

Nice work. Nice clean well structured code. Clear instructions.

File: client/checkout/api/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/api/index.js:147]: @allendav checking out the branch from https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2851 allows us to test this. I didn't have it in the testing instructions, because this is change that I brought over from the classic checkout and I didn't test it deeply with blocks.

File: client/checkout/classic/index.js
Comment (RadoslavGeorgiev) [reply client/checkout/classic/index.js]: Thank you for the suggestion, it's implemented in https://github.com/Automattic/woocommerce-payments/pull/723/commits/70078ba71059638fa86d08b5126104b287896b20.

File: includes/class-wc-payments-payment-method.php
Comment (RadoslavGeorgiev) [reply includes/class-wc-payments-payment-method.php]: I was following the naming scheme from the blocks repo, but your suggestion makes sense. I renamed the class in https://github.com/Automattic/woocommerce-payments/pull/723/commits/ec79d0a1a0343e8db1a160656b7534c60fbcb563.

File: includes/class-wc-payments.php
Comment (RadoslavGeorgiev) [reply includes/class-wc-payments.php]: See https://github.com/Automattic/woocommerce-payments/pull/723#discussion_r494209561

File: includes/class-wc-payment-gateway-wcpay.php
Comment (RadoslavGeorgiev) [reply includes/class-wc-payment-gateway-wcpay.php]: I added `inbox` in https://github.com/Automattic/woocommerce-payments/pull/723/commits/317b4b013306580b466277a9c417edafa49c10ab

**Comment by RadoslavGeorgiev** [comment]
Everything is up-to-date with master and seems to work well, however I cannot figure out why e2e tests are failing in Travis but work well locally. I'll continue with this tomorrow.

**Comment by RadoslavGeorgiev** [comment]
@allendav and @dwainm thank you both for the reviews, and special thanks to @DanReyLop for the `spike/blocks-checkout-merge-with-ie11-fix` branch!

There are a few minor changes/fixes since the last review, so I'd like a last pair of 👀 before merging the PR.

Various things for full blocks support are missing in this PR. I'm creating the https://github.com/Automattic/woocommerce-payments/issues/947 epic to discuss them.

**Comment by allendav** [comment]
Noting this in case anyone else runs into it. Did a `npm run build:release` but the resulting ZIP had a Fatal error:

```Fatal error: Uncaught Error: Class 'MyCLabs\Enum\Enum' not found in /home/dh_yrka4x/staging2.marsiswaiting.com/wp-content/plugins/woocommerce-payments/includes/constants/class-payment-initiated-by.php:20 ```

Did a composer install and then npm run build:release again. That fixed it.

**Comment by allendav** [comment]
Tests nicely. I did notice, however, that when I used the 4000 0027 6000 3184 test card and purposely failed the authentication modal, even after entering a different card, the submit button was still deactivated. Please open a separate issue for this to fix in the next sprint, which authentication is not yet a likely flow

Screen Shot 2020-09-29 at 3 18 03 PM

**Review by allendav — APPROVED** [review]
Tests nicely. Approved. 

Please add a changelog entry for 1.6.0 for this before merging

Please also capture at least following things in the "loose ends" epic #947 

- Test with create account during checkout once that is merged in the blocks repo
- Change confirmCardPayment to async when webpack supports it
- Fix the submit button continuing to be disabled in the event the buyer's card requires and fails authentication and they enter a new card

Nice work. :shipit:

**Comment by RadoslavGeorgiev** [comment]
- [x] Please add a changelog entry for 1.6.0 for this before merging

Please also capture at least following things in the "loose ends" epic #947

- [x] Test with create account during checkout once that is merged in the blocks repo
- [ ] Change confirmCardPayment to async when webpack supports it
- [ ] Fix the submit button continuing to be disabled in the event the buyer's card requires and fails authentication and they enter a new card

**Comment by haszari** [comment]
FYI we've changed the API for registering a payment method in WooCommerce Blocks - now pass an options object (previously was a callback). 

Check this PR for more info: https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3404

I recommend that you update the blocks integration in WCPay as soon as practical - thanks!

**Pull request merged**
